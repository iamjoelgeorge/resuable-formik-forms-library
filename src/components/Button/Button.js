import React from 'react';

import styles from './Button.module.scss';

const Button = (props) => {
  const { label, type, onClick, ...rest } = props;

  return (
    <button className={styles.button} type={type} {...rest}>
      Submit
    </button>
  );
};

export default Button;
