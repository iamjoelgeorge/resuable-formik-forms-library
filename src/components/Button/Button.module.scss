@import '../../theme/variables.scss';

.button {
  border: none;
  padding: 10px;
  border-radius: 2px;
  background-color: $darkPurple;
  color: white;
  font-weight: bold;
  margin-top: 12px;

  &:disabled {
    background-color: rgb(89, 89, 89);
  }
}
