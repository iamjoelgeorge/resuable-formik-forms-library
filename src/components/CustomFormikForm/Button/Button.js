import React from 'react';

import PropTypes from 'prop-types';

import styles from './Button.module.scss';
import { joinClassNames } from '../../../utils/utils';
import {
  ExternalLinkIconWhite,
  ExternalLinkIconRed,
  ExternalLinkIconPurple,
} from '../../../constants/icons';

const Button = (props) => {
  const {
    variant,
    type,
    href,
    label,
    onClick,
    isDisabled,
    theme,
    showExternalLinkIcon,
    containerClass,
    formik,
    ...rest
  } = props;

  // const disableButton = type === 'submit' ? !(formik.dirty && formik.isValid) : isDisabled;
  const disableButton = type === 'submit' ? !formik.isValid : isDisabled;
  const goTo = !isDisabled ? href : '';

  const linkDisabledClass = isDisabled ? styles.linkDisabled : '';
  const linkAsButtonDisabledClass = isDisabled ? styles.linkButtonDisabled : '';

  const IconForLink = theme === 'purple' ? ExternalLinkIconPurple : ExternalLinkIconRed;

  const buttonClasses =
    theme === 'purple'
      ? joinClassNames([styles.button, styles.purpleTheme, containerClass])
      : joinClassNames([styles.button, styles.redTheme, containerClass]);

  const linkClasses =
    theme === 'purple'
      ? joinClassNames([styles.link, styles.purpleTheme, linkDisabledClass, containerClass])
      : joinClassNames([styles.link, styles.redTheme, linkDisabledClass, containerClass]);

  const linkAsButtonClasses = joinClassNames([
    buttonClasses,
    styles.linkAsButton,
    linkAsButtonDisabledClass,
  ]);

  const handleClickOnLink = (e) => {
    if (isDisabled) e.preventDefault();
  };

  const renderButton = () => {
    switch (variant) {
      case 'button':
        return (
          <button className={buttonClasses} type={type} disabled={disableButton} {...rest}>
            {label}
          </button>
        );

      case 'link':
        return (
          <a
            className={linkClasses}
            href={goTo}
            target='_blank'
            onClick={handleClickOnLink}
            rel='noreferrer'
          >
            <span>{label}</span>
            {showExternalLinkIcon && (
              <img className={styles.externalLinkIcon} src={IconForLink} alt='External Link' />
            )}
          </a>
        );

      case 'link_as_button':
        return (
          <a
            className={linkAsButtonClasses}
            href={goTo}
            target='_blank'
            onClick={handleClickOnLink}
            rel='noreferrer'
          >
            {label}
            {showExternalLinkIcon && (
              <img
                className={styles.externalLinkIcon}
                src={ExternalLinkIconWhite}
                alt='External Link'
              />
            )}
          </a>
        );

      default:
        return (
          <button
            className={buttonClasses}
            type={type}
            disabled={disableButton}
            onClick={onClick}
            {...rest}
          >
            {label}{' '}
            {showExternalLinkIcon && (
              <img
                className={styles.externalLinkIcon}
                src={ExternalLinkIconWhite}
                alt='External Link'
              />
            )}
          </button>
        );
    }
  };

  return <>{renderButton()}</>;
};

Button.propTypes = {
  variant: PropTypes.oneOf(['button', 'link', 'link_as_button']),
  type: PropTypes.oneOf(['submit', 'button']),
  href: PropTypes.string,
  label: PropTypes.string,
  onClick: PropTypes.func,
  isDisabled: PropTypes.bool,
  theme: PropTypes.oneOf(['red', 'purple']),
  showExternalLinkIcon: PropTypes.bool,
  containerClass: PropTypes.string,
};

Button.defaultProps = {
  variant: 'button',
  type: 'submit',
  href: '#',
  label: '',
  onClick: () => {},
  isDisabled: false,
  theme: 'purple',
  showExternalLinkIcon: false,
  containerClass: '',
};

export default Button;
