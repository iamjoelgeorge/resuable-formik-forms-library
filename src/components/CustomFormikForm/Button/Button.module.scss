@import '../../../theme/variables.scss';
@import '../../../theme/utils.scss';

* {
  box-sizing: border-box;
}

@mixin useButtonTheme($bgColor, $hoverBgColor, $focusBgColor) {
  background-color: $bgColor;

  &:hover:not(:disabled),
  &:active {
    cursor: pointer;
    background-color: $hoverBgColor;
  }

  &:focus {
    border-color: $focusBgColor;
  }
}

@mixin useLinkTheme(
  $defaultTextColor,
  $hoverColor,
  $borderBottomColorOnFocus,
  $borderBottomColorOnActive,
  $textColorOnActive
) {
  color: $defaultTextColor;

  span {
    border: 2px solid transparent;
    padding-bottom: 2px;
  }

  &:hover:not(&.linkDisabled) {
    color: $hoverColor;
  }

  &:focus {
    span {
      border-bottom-style: dashed;
      border-bottom-color: $borderBottomColorOnFocus;
    }
  }

  &:active {
    border-color: $borderBottomColorOnActive;
    color: $textColorOnActive;
  }
}

.button,
.link {
  border: 3px solid transparent;
  border-radius: 4px;
  font-size: em(16);
  font-weight: $fontWeightExtraBold;
  line-height: 19px;
  margin-top: 12px;
  padding: 0 em(15);
  width: 198px;
  height: 51px;
  text-align: center;
}

.button {
  color: $white;
  transition: 0.1s all ease-in-out;
  outline: none;

  &.purpleTheme {
    @include useButtonTheme($primary, $darkPurple, $brightPurple);
  }

  &.redTheme {
    @include useButtonTheme($tooltipRed, $buttonHover, $pinkOne);
  }

  &:disabled {
    background-color: $white;
    color: $tertiaryGrey;
    border-color: $dGrey;
  }
}

.linkAsButton {
  display: flex;
  justify-content: center;
  align-items: center;
  text-decoration: none;

  &.linkButtonDisabled {
    background-color: $white !important;
    cursor: not-allowed !important;
    color: $tertiaryGrey;
    border-color: $dGrey;
  }
}

.link {
  display: flex;
  justify-content: center;
  align-items: center;
  text-decoration: none;
  outline: none;

  &.purpleTheme {
    @include useLinkTheme($primary, $darkPurple, $darkPurple, $brightPurple, $darkPurple);
  }

  &.redTheme {
    @include useLinkTheme($tooltipRed, $buttonHover, $tooltipRed, $pinkOne, $buttonHover);
  }

  &.linkDisabled {
    color: $tertiaryGrey;
    pointer-events: none;

    .externalLinkIcon {
      -webkit-filter: invert(50%); /* safari 6.0 - 9.0 */
      filter: invert(50%);
    }
  }
}

.externalLinkIcon {
  margin-left: em(8);
}
