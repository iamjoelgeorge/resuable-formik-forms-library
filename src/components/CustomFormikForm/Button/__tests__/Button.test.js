import React from 'react';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';

import Button from '../Button';

const defaultProps = {
  variant: 'button',
  type: 'submit',
  href: '#',
  label: 'submit',
  onClick: () => {},
  isDisabled: false,
  theme: 'purple',
  showExternalLinkIcon: false,
  containerClass: '',
  formik: {
    isValid: true,
  },
};

afterEach(cleanup);
describe('[Component]: Button', () => {
  it('should match the Snapshot', () => {
    const tree = renderer.create(<Button {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the button correctly', () => {
    const props = {
      ...defaultProps,
    };
    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('button-submit')).toHaveTextContent('submit');
  });

  it('should render a disabled button', () => {
    const props = {
      ...defaultProps,
      type: 'button',
      isDisabled: true,
    };
    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('button-submit')).toHaveAttribute('disabled');
  });

  it('should render the link correctly', () => {
    const props = {
      ...defaultProps,
      variant: 'link',
    };
    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link-submit')).toHaveTextContent('submit');
  });

  it('should render "link_as_button" link  correctly', () => {
    const props = {
      ...defaultProps,
      variant: 'link_as_button',
    };
    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link_as_button-submit')).toHaveTextContent('submit');
  });

  it('should render the button with the red theme', () => {
    const props = {
      ...defaultProps,
      theme: 'red',
    };

    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('button-submit')).toHaveClass('redTheme');
  });

  it('should render the link with the red theme', () => {
    const props = {
      ...defaultProps,
      variant: 'link',
      theme: 'red',
    };

    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link-submit')).toHaveClass('redTheme');
  });

  it('should render the "link_as_button" link with the red theme', () => {
    const props = {
      ...defaultProps,
      variant: 'link_as_button',
      theme: 'red',
    };

    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link_as_button-submit')).toHaveClass('redTheme');
  });

  it('should render the button with the purple theme', () => {
    const props = {
      ...defaultProps,
      theme: 'purple',
    };

    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('button-submit')).toHaveClass('purpleTheme');
  });

  it('should render the link with the purple theme', () => {
    const props = {
      ...defaultProps,
      variant: 'link',
      theme: 'purple',
    };

    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link-submit')).toHaveClass('purpleTheme');
  });

  it('should render the "link_as_button" link with the purple theme', () => {
    const props = {
      ...defaultProps,
      variant: 'link_as_button',
      theme: 'purple',
    };

    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link_as_button-submit')).toHaveClass('purpleTheme');
  });

  it('should render the link with the disabled class', () => {
    const props = {
      ...defaultProps,
      variant: 'link',
      isDisabled: true,
    };
    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link-submit')).toHaveClass('linkDisabled');
  });

  it('should render the "link_as_button" link with the disabled class', () => {
    const props = {
      ...defaultProps,
      variant: 'link_as_button',
      isDisabled: true,
    };
    const { getByTestId } = render(<Button {...props} />);
    expect(getByTestId('link_as_button-submit')).toHaveClass('linkButtonDisabled');
  });
});
