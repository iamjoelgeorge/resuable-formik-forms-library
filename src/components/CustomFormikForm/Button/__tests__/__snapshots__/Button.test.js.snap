// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component]: Button should match the Snapshot 1`] = `
<button
  className="button purpleTheme "
  data-testid="button-submit"
  disabled={false}
  type="submit"
>
  submit
</button>
`;
